### ISSUE TITLE
[Question] SoftmaxLikelihood use of its mixing_weights_prior argument

### ISSUE THREAD

**Issue by alexmaraval** [cornellius-gp/gpytorch#1432]
The class `gpytorch.likelihoods.softmax_liklihood.SoftmaxLikelihood` allows for mixing weights (`mixing_weights=True`) and also a prior on those weights (`mixing_weights_prior`).

However, the mixing weights are initialized like so
```
self.register_parameter(
    name="mixing_weights",
    parameter=torch.nn.Parameter(torch.randn(num_classes, num_features).div_(num_features)),
)
```
which means using a random normal. If you pass an instance of a gpytorch prior as `mixing_weights_prior` it is not used to initialize the weights. Is that desired ? Maybe gpytorch priors were not meant to be used in likelihood weights priors but I thought the question was worth asking.

Also, and more importantly, I was wondering how the `mixing_weights_prior` was used in fact ? I understand that it is used in the computation of the `VariationalELBO` in the `log_prior` term. But shouldn't it also be used when computing the `log_likelihood` term ?
For example in the `SoftmaxLikelihood` case, samples are taken from the variational distribution (in `_Likelihood._draw_likelihood_samples()`) and passed to `SoftmaxLikelihood.forward()`. There, they are multiplied with the `mixing_weights` to get the logits of the returned `Categorical` distribution. So the `mixing_weights` are definitely used but I do not see where the `mixing_weights_prior` is being used here.

Should the `mixing_weights` be sampled from the `mixing_weights_prior` somehow ? Otherwise we keep using the initial weights.

Any help/thought would be very much appreciated :) 
Thanks !

**Comment by gpleiss** [comment]
> If you pass an instance of a gpytorch prior as mixing_weights_prior it is not used to initialize the weights. Is that desired ?

That is a good question. At the moment, I don't think that any of our priors affect the initialization of our parameters. It might be worth changing this behavior though.

> Also, and more importantly, I was wondering how the mixing_weights_prior was used in fact ? I understand that it is used in the computation of the VariationalELBO in the log_prior term. But shouldn't it also be used when computing the log_likelihood term ?

It is used. The [`register_prior` line](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/likelihoods/softmax_likelihood.py#L42) adds the prior to the GP model. The VariatioanlELBO class loops through registered priors and adds the prior's log_likelihood to the ELBO.

**Comment by alexmaraval** [comment]
Hi, sorry for the late reply and thanks for yours :)

For the first part of the question, I then modified the `SoftmaxLikelihood` class so it can use the prior distribution on its weights for initialisation.

As for the second part of the question, maybe I asked in a confusing way. I see that the `prior` objects are going to be registered and therefore used later on in the `_ApproximateMarginalLogLikelihood.forward()` computation like so

```
# Log prior term
log_prior = torch.zeros_like(log_likelihood)
for name, module, prior, closure, _ in self.named_priors():
    log_prior.add_(prior.log_prob(closure(module)).sum().div(self.num_data))
```

But in my opinion, if the likelihood is itself parameterized with those parameters having priors, it is different from the original use of the `prior` objects, which are typically used as priors over GP hyperparameters (as far as I understand). So correct me if I'm wrong but that won't be the same in the ELBO if we have priors over parameters of the likelihood, i.e. parameters that do not appear in the GP model at all. The following term in `_ApproximateMarginalLogLikelihood.forward()` computes the expected log likelihood w.r.t. variational distributions.
```
log_likelihood = self._log_likelihood_term(approximate_dist_f, target, **kwargs).div(num_batch)
```
I'd argue that if we treat the Likelihood parameters as variational, then the expectation of the log likelihood should also be w.r.t. to them. 

But I think I'll close this issue for now anyway :)
